akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala

```diff
@@ -169,25 +171,35 @@ object Behaviors {
    * This type of Behavior wraps another Behavior while allowing you to perform
    * some action upon each received message or signal. It is most commonly used
    * for logging or tracing what a certain Actor does.
+   *
+   * @param id Used to identify when a tap is added again (using instance equality),
+   *           and the outermost/previous one should be removed. If not called recursively,
+   *           just pass a new `InterceptId` instance.
    */
   def tap[T](
     clazz:     Class[T],
     onMessage: Procedure2[ActorContext[T], T],
     onSignal:  Procedure2[ActorContext[T], Signal],
+    id:        WrappedBehaviorId,
```

Comment:
pretty ugly that we have to expose this all the way out to public api

Reply:
I wonder if we should start think about a more general intercept API instead of this `tap`.
Could an interceptor api look like this?

```scala
trait BehaviorInterceptor[T] {
  def aroundReceive(ctx: ActorContext[T], msg: T, target: ReceiveTarget[T]): Behavior[T]
  def aroundSignal(ctx: ActorContext[T], signal: Signal, target: SignalTarget[T]): Behavior[T]
}

trait ReceiveTarget[T] {
  def apply(msg: T): Behavior[T]
}

trait SignalTarget[T] {
  def apply(signal: Signal): Behavior[T]
}
```

I think that covers `tap` and also our internal `intercept`.

As a bonus we can then use equals of the `BehaviorInterceptor` as the `WrappedBehaviorId`, and it doesn't have to be exposed to public.

Reply:
and it would cover https://github.com/akka/akka/pull/25508

Reply:
I like the idea about a general intercept API, but i'm not quite convinced it helps with the deduplication of nested behaviors (unless we completely discard the idea of factories that takes lambdas for decoration and always require the user to subclass which means a lot more boilerplate). 

For the record I agree that the identifier is ugly, but I couldn't figure out another way to force the user of the wrapping behavior factory to carefully think about the problem.

Reply:
(I'll explore that idea even though sceptic)

Reply:
I think `BehaviorInterceptor` can replace tap, i.e. no lambdas there anymore.

For `monitor` we have the ActorRef as the unique identifier.

For `withTimers` we know there is only one.

What other things do we have? `widen`?

Reply:
I think _any_ nesting behavior should be done through interception so that would be supervise as well.

Wen't with those APIs exactly as in the idea, has been looking good so far, but now I got stuck on implementing withTimers since that can also filter out messages - with the `MessageTarget` we don't have a behavior to return for that. So I will have to revise that somehow.

Reply:
Went with an additional `MessageTarget.ignore`

Reply:
@johanandren isn't it just returning `same` without calling `MessageTarget`?

Reply:
Ah, ofc